### PR TITLE
feature 32960 - Enabled document downloads using streams

### PR DESCRIPTION
**Pull request by Rick-Ritense**
Enabled rest methods for this application

Created a rest endpoint to download documents via streams

### CODE REVIEW

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/service/DownloadService.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import org.springframework.web.bind.annotation.GetMapping
+import org.springframework.web.bind.annotation.PathVariable
+import org.springframework.web.bind.annotation.RequestMapping
+import org.springframework.web.bind.annotation.RestController
+import java.util.UUID
+import javax.servlet.http.HttpServletResponse
+
+@RestController
+@RequestMapping( value = ["/api"])
+interface DownloadService {
```
Comment (Klaas-Ritense) [inline zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/service/DownloadService.kt:27]: imo this should be named `DownloadResource` and should be in the `.web.rest` package

The other DownloadResource should then be in the `.web.rest.impl` package

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt [hunk 2]
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2015-2021 Ritense BV, the Netherlands.
```
Comment (Klaas-Ritense) [inline zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt:2]: 2022 :p

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    override fun downloadStreaming(response: HttpServletResponse, documentId: UUID) {
+        response.setHeader("Content-Disposition", "attachment")
+
+        // Request service to get the file's data stream
+        val fileDataStream = openZaakClient.getDocumentContentStream(documentId)
+
+        // Stream to the response's output stream instead of loading it all in memory
+        DataBufferUtils.write(fileDataStream, response.outputStream)
+            .map(DataBufferUtils::release)
+            .blockLast()
```
Comment (Klaas-Ritense) [inline zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt:37]: Maybe its needed to set the `Content-Type` also. But Im not sure

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/autoconfiguration/DocumentAutoConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    @Bean
+    @ConditionalOnMissingBean(DocumentResourceHttpSecurityConfigurer::class)
+    fun documentResourceHttpSecurityConfigurer(): DocumentResourceHttpSecurityConfigurer {
+        return DocumentResourceHttpSecurityConfigurer()
+    }
+
+    @Bean
+    @ConditionalOnMissingBean(DownloadResource::class)
+    fun documentResource(openZaakClient: OpenZaakClient): DownloadResource {
```
Comment (Klaas-Ritense) [inline zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/autoconfiguration/DocumentAutoConfiguration.kt:39]: Typo. Should be `downloadResource`

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import org.springframework.core.io.buffer.DataBufferUtils
+import java.util.UUID
+import javax.servlet.http.HttpServletResponse
+
+class DownloadResource(
+    val openZaakClient: OpenZaakClient
+) : DownloadService {
+
+    override fun downloadStreaming(response: HttpServletResponse, documentId: UUID) {
+        response.setHeader("Content-Disposition", "attachment")
```
Comment (ivo-ritense) [inline zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt:29]: Does this work properly in browsers without including the file name? What will be the suggested name of the file of the download like this?

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt (hunk 2 above)
Comment (Rick-Ritense) [reply zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt:2]: De licenseHeaderFile.template die in die package staat toont nog het jaartal 2021. Die file wordt gebruikt voor de controle dat de license tekst klopt. 

Daarnaast is het t/m jaartal niet belangrijk voor open source licenties

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/autoconfiguration/DocumentAutoConfiguration.kt (hunk 4 above)
Comment (Rick-Ritense) [reply zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/autoconfiguration/DocumentAutoConfiguration.kt:39]: Yes you are right. I renamed the file, forgot about this method afterwards :)

**Review by ivo-ritense — CHANGES_REQUESTED** [review]
In general it looks like the biggest issue is the lack of security as far as i can tell. We need to make sure that the document has a link to a zaak that the user is related to.

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/security/config/DocumentResourceHttpSecurityConfigurer.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.http.HttpMethod.GET
+import org.springframework.security.config.annotation.web.builders.HttpSecurity
+
+@Configuration
+class DocumentResourceHttpSecurityConfigurer {
+
+    fun configure(http: HttpSecurity) {
+        try {
+            http.authorizeRequests()
+                .antMatchers(GET, "/api/document/{documentId}/content").permitAll()
```
[the remark this answers is not in the document]
Comment (Rick-Ritense) [reply zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/security/config/DocumentResourceHttpSecurityConfigurer.kt:29]: I will change it

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/client/OpenZaakClient.kt [hunk 7]
```diff
@@ -157,7 +159,16 @@ class OpenZaakClient(
...
+    fun getDocumentContentStream(id: UUID): Flux<DataBuffer> {
+        return webClient()
+            .get()
+            .uri("/documenten/api/v1/enkelvoudiginformatieobjecten/$id/download")
+            .accept(MediaType.APPLICATION_OCTET_STREAM)
+            .retrieve()
+            .bodyToFlux(DataBuffer::class.java)
+    }
+
+    fun webClient(): WebClient {
```
[the remark this answers is not in the document]
Comment (Rick-Ritense) [reply zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/client/OpenZaakClient.kt:171]: Before the getDocumentContentStream() was in another file That extended this class.
I forgot to revert it back to private afterwards

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/service/DownloadService.kt (hunk 1 above)
Comment (Rick-Ritense) [reply zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/service/DownloadService.kt:27]: I checked and that is the same setup as in valtimo-backend-libraries indeed. I will change it

File: zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt (hunk 5 above)
Comment (Rick-Ritense) [reply zaak/src/main/kotlin/com/ritense/valtimo/portal/zaak/web/rest/DownloadResource.kt:29]: Yes this should work according to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Disposition

When a file name is provided it is used for a "Save As" dialog. But I can't be sure what type is document is being downloaded. That is why I left the filename blank

**Comment by ivo-ritense** [comment]
> In general it looks like the biggest issue is the lack of security as far as i can tell. We need to make sure that the document has a link to a zaak that the user is related to.

After talking about it we agreed that the underlying document API is reponsible according to the design. Only think that is needed here is to ensure the user is authenticated.

**Review by ivo-ritense — APPROVED** [review]

**Review by ivo-ritense — APPROVED** [review]

**Pull request merged**